### DISCUSSION TITLE
How to get another page object in a page?

### DISCUSSION THREAD

**Discussion by Mister-Hope** [vuepress/core#228] [Q&A] [answered]
E.g.: Just expect I am under path `/` which is `readme.md`. how can I get `/example.md` (`/example.html` or even rendered to another link when `premallink` is set)?

The provided `usePagesData()` is in a format of `Record<PageKey, ()=>Promise<PageData>>`, which I think is the correct composition api I should use, but I must access a specfic page through page key (in format `v-xxxxx`).

I have no idea to get a page key from a path ( I am not finding any related api exported).

Update:

![image](https://user-images.githubusercontent.com/33315834/122712251-33bec600-d296-11eb-958a-f9e3b8309289.png)

Shall I hash the path myself? I think we should better provide an api. @meteorlxy As I won't find any pages if users are using premallinks.

![image](https://user-images.githubusercontent.com/33315834/122712436-7f716f80-d296-11eb-80a9-d60cc7c71505.png)

### Why I need this.

This need is required for me to provide similar features like `displayAllHeaders` in v1, where the input are just "infferPathI" intead of "permallinks" configured by users.

**Comment by meteorlxy** [comment, marked as answer]
The `key` is the name property of vue-router. Try using this to get the routes: https://next.router.vuejs.org/api/#getroutes

BTW, `usePagesData` is also an undocumented API

**Reply by Mister-Hope** [reply]
Yes, I know that, I think we should at least give developer ability to visit any PageData they need to instead of only current page.

**Reply by Mister-Hope** [reply]
Any plans to add this in docs?(Also I think we should doc that how to get page keys)

**Reply by meteorlxy** [reply]
Page data may only belong to the corresponding page. I think we'll inject page data into page component as [vitepress does](https://github.com/vuejs/vitepress/blob/c046905b032a765352ff6bb9944f72db76c5cf45/src/node/markdownToVue.ts#L90-L92). Then the `usePagesData` will be removed.

You can consider collecting data of all pages on demand in build process, and write a temp file for that.

**Reply by Mister-Hope** [reply]
Then how can a reactive plugin-blog be possible? I do think making temp files in build process to provide blog feature is possible, but I don't think it can run on devServer without a build(not even none-reactive)

Or in other words to say, vuepress2 will only focus on docs? 

Up to you, if so, I will fork and make a vuepress-blog package, and in that case, I still think vuepress is weak comparing to other docs builders when both in "out-of-box". An official support of features like seo and sitemap should be added. I can make prs for that.

**Reply by meteorlxy** [reply]
What do you mean by "reactive" here? I don't understand why things could not be "reactive".

> I will fork and make a vuepress-blog package

Always feel free to make your own fork. It's allowed by MIT license. No need to make a declaration here.

> I can make prs for that

We are always accepting PRs.

**Reply by Mister-Hope** [reply]
In the plugin api you are providing now, I didnot see life cycle like `onPageUpdate`, and in early versions, I don't think `extendsPageData` will be called when users are changing frontmatters. So how can the temp files being updated(I supposed I am collecting it through `extendsPageData`) when users change frontmatter (e.g: setting `tag` to another value)?

I try to build a simple blog plugin in the last few alpha versions. But I can not even start an idea. As the pages array was available in v1.

**Reply by Mister-Hope** [reply]
> We are always accepting PRs.

Then I am happy to move my seo and sitemap plugin under `@vuepress` after your review. These two are absolutely powerfull then current plugins like `vuepress-plugin-seo` and `vuepress-plugin-sitemap`.

**Reply by Mister-Hope** [reply]
Oh, BTW, if I understands you well, I must prepare some temp files on my own collecting all pages data if I want to access another pageData in one page? I still think it may be common in themes to access other pages data. E.g:

- BreadCrumb
- Sidebar providing other page headers
- Some I18n features which need to access the alternate language pages.
- Some analyze page just like blog pages, e.g: a sitemap page
- 
So I still think we can make a plugin to generate temp files for people to do this.

**Reply by meteorlxy** [reply]
> I don't think extendsPageData will be called when users are changing frontmatters

If you find something does not work as expected, create a bug report.
